File: M3Tools/Dialogs/Google/UploadFileDialog.vb
﻿Imports System.Collections.ObjectModel
Imports System.ComponentModel
Imports System.Windows.Forms

Public Class UploadFileDialog
	'Private ReadOnly __files As New GTools.Types.FileCollection
	Private __permission As Global.Google.Apis.Drive.v3.Data.Permission
	Public ReadOnly Property Files As IList 'GTools.Types.FileCollection
		Get
			Return bsFiles.List '__files
		End Get
		'Set(value As GTools.Types.FileCollection)
		'	__files.Clear()
		'	__files.AddRange(value)
		'End Set
	End Property

	Sub Loading(sender As Object, e As EventArgs) Handles Me.Load
		dt_DriveHeirarchy.Reload()
	End Sub

	Private Sub Upload(sender As Object, e As EventArgs) Handles btn_Upload.Click
		If dt_DriveHeirarchy.SelectedNode IsNot Nothing Then
			If chk_DefaultPermissions.Checked Then
				__permission = Nothing
			Else
				Using permissions As New PermissionsDialog()
					If permissions.ShowDialog() = DialogResult.OK Then
						Console.WriteLine(permissions.Permission.Role)
						Console.WriteLine(permissions.Permission.Type)
						__permission = permissions.Permission
					End If
				End Using
			End If

			'bw_LoadFiles.RunWorkerAsync()
		Else
			MessageBox.Show("You must select a parent folder.", "New Upload", MessageBoxButtons.OK, MessageBoxIcon.Error)
		End If
	End Sub

	Private Sub Cancel(sender As Object, e As EventArgs) Handles Cancel_Button.Click
		Me.DialogResult = DialogResult.Cancel
		Me.Close()
	End Sub

	Private Sub DialogLoad(sender As Object, e As DoWorkEventArgs) Handles bw_LoadDialog.DoWork

	End Sub

	Private Sub DialogLoadFinished(sender As Object, e As RunWorkerCompletedEventArgs) Handles bw_LoadDialog.RunWorkerCompleted
		dt_DriveHeirarchy.Reload()
	End Sub

	Private Sub Finish()
		Me.DialogResult = DialogResult.OK
		Me.Close()
	End Sub

	Private Sub LoadFiles(sender As Object, e As DoWorkEventArgs) Handles bw_LoadFiles.DoWork
		Dim fileParts() As String
		Dim fileExt As String
		Files.Clear()
		For Each file As Types.GTools.File In bsFiles.List 'fu_FileUpload.Files
			fileParts = file.Name.Split("."c)
			fileExt = fileParts(fileParts.Length - 1)
			Files.Add(New Types.GTools.File("", file.Name, fileExt))
		Next
	End Sub

	Private Sub SetParents(sender As Object, e As RunWorkerCompletedEventArgs) Handles bw_LoadFiles.RunWorkerCompleted
		Dim selectedParentId = dt_DriveHeirarchy.SelectedNode.Name
		For Each file As Types.GTools.File In Files
			If file.Parents Is Nothing Then
				file.Parents = If(selectedParentId.Equals("main"), Nothing, New Collection(Of String)({selectedParentId}))
			Else
				If selectedParentId.Equals("main") Then
					file.Parents = Nothing
				Else
					file.Parents.Add(dt_DriveHeirarchy.SelectedNode.Name)
				End If
			End If
		Next
		bw_UploadFiles.RunWorkerAsync()
	End Sub

	Private Sub UploadFiles(sender As Object, e As DoWorkEventArgs) Handles bw_UploadFiles.DoWork
		Dim fileName As String
		Try
			For Each file As Types.GTools.File In Files
				fileName = InputBox($"What would you like to call this file? (default: {Utils.DefaultFileName(file.Name)})", "File Name", Utils.DefaultFileName(file.Name))
				gdt_GDrive.UploadFile(file, fileName, __permission)
			Next

			e.Result = True
			Finish()
		Catch ex As Exceptions.UploadException
			MessageBox.Show(Me, ex.Message, "New Upload", MessageBoxButtons.OK, MessageBoxIcon.Error)
			e.Result = False
		End Try
	End Sub

	Private Sub UploadCompleted(sender As Object, e As RunWorkerCompletedEventArgs) Handles bw_UploadFiles.RunWorkerCompleted
		If CBool(e.Result) Then
			Finish()
		End If
	End Sub
End Class

File: M3Tools/Dialogs/Google/UploadFileDialog.cs
using System;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;

using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

// TODO: Clean this up better
namespace SPPBC.M3Tools
{
	/// <summary>
	/// 
	/// </summary>
	public partial class UploadFileDialog
	{
		// Private ReadOnly __files As New GTools.Types.FileCollection
		private Google.Apis.Drive.v3.Data.Permission __permission;

		/// <summary>
		/// The list of files to be uploaded
		/// </summary>
		public IList Files // GTools.Types.FileCollection
=> bsFiles.List; // __files

		/// <summary>
		/// 
		/// </summary>
		public UploadFileDialog() => InitializeComponent();

		/// <summary>
		/// 
		/// </summary>
		/// <param name="sender"></param>
		/// <param name="e"></param>
		public void Loading(object sender, EventArgs e) => dt_DriveHeirarchy.Reload(this, e);

		private void Upload(object sender, EventArgs e)
		{
			if (dt_DriveHeirarchy.SelectedNode == null)
			{
				_ = MessageBox.Show("You must select a parent folder.", "New Upload", MessageBoxButtons.OK, MessageBoxIcon.Error);
				return;
			}

			if (chk_DefaultPermissions.Checked)
			{
				__permission = null;
			}
			else
			{
				using PermissionsDialog permissions = new();

				if (permissions.ShowDialog() != DialogResult.OK)
				{
					return;
				}

				Console.WriteLine(permissions.Permission.Role);
				Console.WriteLine(permissions.Permission.Type);
				__permission = permissions.Permission;
			}

			// bw_LoadFiles.RunWorkerAsync()
		}

		private void Cancel(object sender, EventArgs e)
		{
			DialogResult = DialogResult.Cancel;
			Close();
		}

		private void DialogLoad(object sender, DoWorkEventArgs e)
		{

		}

		private void DialogLoadFinished(object sender, RunWorkerCompletedEventArgs e) => dt_DriveHeirarchy.Reload(this, e);

		private void Finish()
		{
			DialogResult = DialogResult.OK;
			Close();
		}

		private void LoadFiles(object sender, DoWorkEventArgs e)
		{
			string[] fileParts;
			string fileExt;
			Files.Clear();
			foreach (Types.GTools.File @file in bsFiles.List) // fu_FileUpload.Files
			{
				fileParts = @file.Name.Split('.');
				fileExt = fileParts[fileParts.Length - 1];
				_ = Files.Add(new Types.GTools.File("", @file.Name, fileExt));
			}
		}

		private void SetParents(object sender, RunWorkerCompletedEventArgs e)
		{
			string selectedParentId = dt_DriveHeirarchy.SelectedNode.Name;
			foreach (Types.GTools.File @file in Files)
			{
				if (@file.Parents is null)
				{
					@file.Parents = selectedParentId.Equals("main") ? null : new string[] { selectedParentId };
				}
				else if (selectedParentId.Equals("main"))
				{
					@file.Parents = null;
				}
				else
				{
					@file.Parents.Add(dt_DriveHeirarchy.SelectedNode.Name);
				}
			}

			bw_UploadFiles.RunWorkerAsync();
		}

		private async void UploadFiles(object sender, DoWorkEventArgs e)
		{
			string fileName;
			try
			{
				foreach (Types.GTools.File @file in Files)
				{
					fileName = Interaction.InputBox($"What would you like to call this file? (default: {Utils.DefaultFileName(@file.Name)})", "File Name", Utils.DefaultFileName(@file.Name));
					_ = await gdt_GDrive.UploadFile(@file, fileName, __permission);
				}

				e.Result = true;
				Finish();
			}
			catch (Exceptions.UploadException ex)
			{
				_ = MessageBox.Show(this, ex.Message, "New Upload", MessageBoxButtons.OK, MessageBoxIcon.Error);
				e.Result = false;
			}
		}

		private void UploadCompleted(object sender, RunWorkerCompletedEventArgs e)
		{
			if (Conversions.ToBoolean(e.Result))
			{
				Finish();
			}
		}
	}
}
